### PR TITLE
Resolution Session Rework

### PR DESCRIPTION
**Pull request by thephoenixofthevoid**
I think it's reasonable to do changes in small steps. In this batch:

1. describeInjection is now a free function. It's was not related to ResolutionSession. Having it as a free function both looks correct and much more convenient. 
2. debugModule => debugFactory, it is just to unify names across the package. In other files debugFactory is used.
3. Also discovered this grouping:
```ts
    if (debug.enabled) {
      debug('Exit binding:', binding?.toJSON());
      debug('Resolution path:', this.getResolutionPath() || '<empty>');
    }
```

## Checklist

- [+] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [+] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [+] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: examples/context/src/custom-inject-resolve.ts [hunk 1]
```diff
@@ -24,7 +25,7 @@ const resolve: ResolverFunction = (
   session: ResolutionSession,
 ) => {
   console.log('Context: %s Binding: %s', ctx.name, session.currentBinding!.key);
-  const targetName = ResolutionSession.describeInjection(injection).targetName;
+  const targetName = describeInjection(injection).targetName;
```
Comment (raymondfeng) [inline examples/context/src/custom-inject-resolve.ts:28]: This is a breaking change from API perspective. Is it really worth the effort? If so, let's keep `ResolutionSession.describeInjection()`.

Comment (thephoenixofthevoid) [reply examples/context/src/custom-inject-resolve.ts:28]: As a quick reply, We may keep it using (there is a way to do it in a typescript way)
```
ResolutionSession.describeInjection = describeInjection
```
if backward compatibility is the main concern. There is a huge cognitive load this function being the static function brings.
I will update this message with more thorough arguments very soon. 
In short: 
1) It is unrelated to the class it is defined on (it doesn't contain a single instance of that class)
2) Its implementation also doesn't use other static functions
3) It breaks symmetry inside ResolutionSession class, which I am trying to expose to use it as a base for refactoring
4) It doesn't seem to be used much outside core (but I might be wrong here -- please correct me if I am wrong)

The refactoring mentioned in (3) seems to be completely equivalent to the current implementation 
in cases where it applies, but allows for parallel resolution, avoiding forks, a more clear mental model, and fewer moving parts that can break locally (ie: push or pop additional time and the entire stack is messed up) and ruin the whole resolution.

Comment (thephoenixofthevoid) [reply examples/context/src/custom-inject-resolve.ts:28]: Another way to preserve compatibility is just have a free function and single line static method that simply calls it. Possibly deprecate it in favour of free function when semantic versioning allows it. Sounds good?

**Comment by thephoenixofthevoid** [comment]
Some checks were not successful with
```
Error: The operation was canceled.
```
Should I do anything about it?

File: packages/context/src/resolution-session.ts [hunk 2]
```diff
@@ -293,12 +276,43 @@ export class ResolutionSession {
   toString() {
     return this.getResolutionPath();
   }
+
+  /**
+   * Describe the injection for debugging purpose
+   *
+   * @TODO remove in next major version
```
Comment (raymondfeng) [inline packages/context/src/resolution-session.ts:283]: Use `@deprecated`?

Comment (raymondfeng) [reply packages/context/src/resolution-session.ts:283]: Deprecation does not count as breaking. It's signaling an upcoming breaking change.

Comment (thephoenixofthevoid) [reply packages/context/src/resolution-session.ts:283]: Also, there is a clear need to draw a line between:
1) End-User APIs -- to compose components and configs into system and eventually run it
2) Extension Author APIs -- Exposes API at the extension point in a coherent way
3) Low-level API  -- Stable API used to do low-lever things in the core (considered public API and under semver)
4) Internal API  -- It's exported but not considered stable, may change at any moment and must not be used outside of monorepo

Your thoughts on that?

At the moment, exported === public, which is incorrect. For example, when a new API surface is being created 
(by splitting a module into two, for example), we at first would export everything that crosses this API surface, 
and work out actual public API of the new module afterwards. But each time we use export we risk it to be 
treated as public API.

Long story short: exported functions should not be automatically treated as public. Public API should be designed with a great care. To me, it seems it already got to the point where inconsistent public API design hinders the development, 
and at the same time prevents from reaching any notion of stability (Long Term Support).

Comment (thephoenixofthevoid) [reply packages/context/src/resolution-session.ts:283]: I apologize if I have been too direct in the last message. I think it's an important piece of information (I looked through all discussions where the consensus has not been reached and tried to figure out the root cause so it can be fixed)

Comment (thephoenixofthevoid) [reply packages/context/src/resolution-session.ts:283]: It's easier than It sounded. Check this thing out: https://api-extractor.com/
It's created by MS to handle this exact kind of problem. 
What do you think on that?

Comment (thephoenixofthevoid) [reply packages/context/src/resolution-session.ts]: My Internal API actually is further divided into: 

- https://api-extractor.com/pages/tsdoc/tag_alpha/
- https://api-extractor.com/pages/tsdoc/tag_beta/
- https://api-extractor.com/pages/tsdoc/tag_internal/

That's looks even better

**Comment by thephoenixofthevoid** [comment]
Is there any progress in reviewing changes?

File: packages/context/src/resolution-session.ts (hunk 2 above)
Comment (achrinza) [reply packages/context/src/resolution-session.ts]: > At the moment, exported === public, which is incorrect

Unfortunately that's not the case with LoopBack 4; `@public` is not used in most of its APIs, hence, IMO, marking it as a non-breaking change would not be in the best interest of the framework users.

**However**, #7533 is a breaking change, and we could drop a second PR to remove it completely.

File: packages/context/src/resolution-session.ts [hunk 3]
```diff
@@ -293,12 +276,42 @@ export class ResolutionSession {
...
+   * @deprecated
+   * Describe the injection for debugging purpose
+   *
+   * @param injection - Injection object
+   * @returns {InjectionDescriptor}
+   */
+  static describeInjection(injection: Readonly<Injection>) {
+    return describeInjection(injection);
+  }
+}
```
Comment (achrinza) [inline packages/context/src/resolution-session.ts:290]:
```suggestion
  /**
   * {@inheritDoc describeInjection}
   * @deprecated Use {@link describeInjection} instead.
   */
  static describeInjection(injection: Readonly<Injection>) {
    return describeInjection(injection);
  }
}
```

File: packages/context/src/resolution-session.ts [hunk 4]
```diff
@@ -215,23 +188,33 @@ export class ResolutionSession {
...
 
-    if (this.stack.find(i => isBinding(i) && i.value === binding)) {
+  /**
+   * @WIP
+   *
+   * @TODO get rid of an argument by checking
+   *       set.size === array.length instead
+   *       after binding is pushed
+   */
+  private checkForCircularDependency(binding: Readonly<Binding>) {
```
Comment (achrinza) [inline packages/context/src/resolution-session.ts:208]:
```suggestion
  /**
   * @internalRemarks
   * #TODO: get rid of an argument by checking
   *        set.size === array.length instead
   *        after binding is pushed
   *
   * @experimental
   */
  private checkForCircularDependency(binding: Readonly<Binding>) {
```

File: packages/context/src/resolution-session.ts [hunk 5]
```diff
@@ -369,46 +382,26 @@ export interface ResolutionContext<T = unknown> {
...
-  }
-
   /**
    * Build the error message for the resolution to include more contextual data
    * @param reason - Cause of the error
-   * @param resolutionCtx - Resolution context
+   * @param ctx - Resolution context
    */
-  private static buildMessage(
-    reason: string,
```
Comment (achrinza) [inline packages/context/src/resolution-session.ts:394]: Was the `reason` parameter removed intentionally?

Otherwise, this might constitute a breaking change as the class isn't sealed.

**Review by achrinza — COMMENTED** [review]
Apologies for the delayed review; I've added my comments above.

It may be a good idea to bundle breaking changes with #7533

File: packages/context/src/resolution-session.ts (hunk 5 above)
Comment (thephoenixofthevoid) [reply packages/context/src/resolution-session.ts:394]: It is not parameter being removed, but the whole private method.

**Comment by thephoenixofthevoid** [comment]
But there are no breaking changes left in this PR. All what is being removed or altered have been explicitly marked as private.
